### PR TITLE
dev-libs/leveldb: bump to 1.20, sub-slot to avoid ABI breakage 

### PR DESCRIPTION
**Pull request by sbraz**
Hi,
This bumps leveldb to 1.20. Because of an ABI change without a SONAME change, I had to introduce a sub-slot. I made sure that all reverse dependencies have the correct slot operator. I changed the stable ceph and qtwebkit straight-to-stable. Other affected ebuilds were unstable.

Changes to the ebuild:
```diff
--- leveldb-1.18-r2.ebuild	2017-11-19 00:05:14.641597450 +0100
+++ leveldb-1.20.ebuild	2018-06-16 02:28:09.808573354 +0200
@@ -1,40 +1,35 @@
-# Copyright 1999-2017 Gentoo Foundation
+# Copyright 1999-2018 Gentoo Foundation
 # Distributed under the terms of the GNU General Public License v2
 
-EAPI=5
+EAPI=7
 
-inherit eutils multilib toolchain-funcs versionator
+inherit multilib toolchain-funcs
 
 DESCRIPTION="a fast key-value storage library written at Google"
 HOMEPAGE="http://leveldb.org/ https://github.com/google/leveldb"
 SRC_URI="https://github.com/google/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz"
 
 LICENSE="BSD"
-SLOT="0"
-KEYWORDS="amd64 arm ~arm64 ~mips ppc ppc64 x86 ~amd64-fbsd ~amd64-linux ~x86-linux"
-IUSE="+snappy static-libs +tcmalloc kernel_FreeBSD"
+# https://github.com/google/leveldb/issues/536
+SLOT="0/1"
+KEYWORDS="~amd64 ~arm ~arm64 ~mips ~ppc ~ppc64 ~x86 ~amd64-fbsd ~amd64-linux ~x86-linux"
+IUSE="+snappy static-libs kernel_FreeBSD +tcmalloc test"
 
 DEPEND="tcmalloc? ( dev-util/google-perftools )
 	snappy? (
 		app-arch/snappy:=
-		static-libs? ( app-arch/snappy[static-libs] )
 	)"
 RDEPEND="${DEPEND}"
 
-src_prepare() {
-	epatch "${FILESDIR}"/${PN}-1.18-mips.patch
-	epatch "${FILESDIR}"/${PN}-1.18-configure.patch #541186
-
-	local SHARED_MINOR=$(get_version_component_range 2)
-	sed \
-		-e "s/\(^ SHARED_MINOR =\).*/\1 ${SHARED_MINOR}/" \
-		"${FILESDIR}/${PN}-1.9.0-memenv-so.patch" > memenv-so.patch
-	epatch memenv-so.patch
-}
+# https://bugs.gentoo.org/651604
+REQUIRED_USE="snappy? ( !static-libs )"
+
+# https://github.com/google/leveldb/issues/234
+# https://github.com/google/leveldb/issues/236
+PATCHES=( "${FILESDIR}"/{${PN}-1.18-configure.patch,${P}-memenv-so.patch} )
 
 src_configure() {
 	# These vars all get picked up by build_detect_platform
-	# which the Makefile runs for us automatically.
 	tc-export AR CC CXX
 	export OPT="-DNDEBUG ${CPPFLAGS}"
 	local targetos
@@ -48,11 +43,13 @@
 	USE_SNAPPY=$(usex snappy) \
 	USE_TCMALLOC=no \
 	TMPDIR=${T} \
-	sh -x ./build_detect_platform build_config.mk ./
+	sh -x ./build_detect_platform build_config.mk ./ || die
 }
 
 src_compile() {
-	emake $(usex static-libs 'libmemenv.a' 'LIBRARY=') all libmemenv.SHARED
+	default
+	usex static-libs && emake out-static/lib{leveldb,memenv}.a
+	use test && emake static_programs
 }
 
 src_test() {
@@ -61,12 +58,12 @@
 
 src_install() {
 	insinto /usr/include
-	doins -r include/*
-	# This matches the path Debian picked.  Upstream provides no guidance.
+	doins -r include/.
+	# This matches the path Debian picked. Upstream provides no guidance.
 	insinto /usr/include/leveldb/helpers
 	doins helpers/memenv/memenv.h
 
-	dolib.so libleveldb*$(get_libname)*
-	use static-libs && dolib.a libleveldb.a libmemenv.a
-	dolib.so libmemenv*$(get_libname)*
+	dolib.so out-shared/libleveldb*$(get_libname)*
+	use static-libs && dolib.a out-static/lib{leveldb,memenv}.a
+	dolib.so out-shared/libmemenv*$(get_libname)*
 }
```

I've also removed a bunch of older unstable versions which aren't required by anything.

@luke-jr Once this is merged and you know leveldb 1.20 works with bitcoin, I think one of the options would be to add slot operators for virtual/bitcoin-leveldb to bitcoind and bitcoin-qt. Then, you would create a new version of the virtual that uses a new sub-slot and only corresponds to leveldb versions that have the same sub-slot, maybe something like :
```sh
SLOT="0/1"
RDEPEND="=dev-libs/leveldb-1.20:0/1"
```

### CODE REVIEW

**Comment by sbraz** [comment]
Ping @patricklauer

**Comment by luke-jr** [comment]
Might make more sense to just have bitcoin stuff depend on the virtual *and* `dev-libs/leveldb:=` ?

File: dev-libs/leveldb/leveldb-1.20.ebuild [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+src_configure() {
+	# These vars all get picked up by build_detect_platform
+	tc-export AR CC CXX
+	export OPT="-DNDEBUG ${CPPFLAGS}"
+	local targetos
+	if use kernel_FreeBSD; then
+		targetos="FreeBSD"
+	else
+		targetos="Linux"
+	fi
```
Comment (SoapGentoo) [inline dev-libs/leveldb/leveldb-1.20.ebuild]: `local targetos="$(usex kernel_FreeBSD FreeBSD Linux)"`

File: dev-libs/leveldb/leveldb-1.20.ebuild [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	usex static-libs && emake out-static/lib{leveldb,memenv}.a
+	use test && emake static_programs
+}
+
+src_test() {
+	emake check
+}
+
+src_install() {
+	insinto /usr/include
```
Comment (SoapGentoo) [inline dev-libs/leveldb/leveldb-1.20.ebuild]: this is kinda awful, but ok...

**Comment by luke-jr** [comment]
Actually, sub-slots don't seem sufficient for this...? Stuff linking to it will break between when the update is installed, and when the rebuild finishes..

**Comment by luke-jr** [comment]
(whereas with a proper soname bump, preserved libs will keep them working in the interim)

**Comment by SoapGentoo** [comment]
@luke-jr 
1) preserved-libs is a portage feature, not a PMS feature
2) preserved-libs and subslots are complementary and not mutually exclusive
3) using portage as package manager, nothing should break

**Comment by luke-jr** [comment]
@SoapGentoo The ABI is broken without a soname bump, so this will *replace* the current library. Replaced libraries cannot be preserved, and will be silently linked against by software built against the old version.

**Pull request closed without merging**